Additional invite/promotion state and missing C functions

mpretty-cyro opened the pull request:
- Added an 'INVITE_NOT_SENT' status which is the default used when creating a new group member
- Added a groups_keys_free function to the C API
- Added a groups_keys_storage_namespace function to the C API
- Updated the promotion functions based on the new status options

mpretty-cyro committed:
Additional invite/promotion state and missing C functions

• Added an 'INVITE_NOT_SENT' status which is the default used when creating a new group memeber
• Added a groups_keys_free function to the C API
• Added a groups_keys_storage_namespace function to the C API
• Updated the promotion functions based on the new status options

simophin commented on src/config/groups/keys.cpp:
```diff
@@ -1432,6 +1432,14 @@ LIBSESSION_C_API int groups_keys_init(
 }
 
+LIBSESSION_C_API void groups_keys_free(config_group_keys* conf) {
+    delete conf;
```
Is it enough to just delete the `conf`? `conf->internals` is a raw pointer it seems.

mpretty-cyro replied:
I just replicated the existing `config_free` function 😬 
```
LIBSESSION_EXPORT void config_free(config_object* conf) {
    delete conf;
}
```

simophin replied:
Hmm I think you do need to delete the internals here because
1. `config_object`/`config_groupb_keys` is a plain struct (it doesn't have a destructor)
2. `internal` is created using `new` and put into a raw pointer, that means it doesn't have a smart pointer to do the clean up.

Unless I miss something I think this one will leak memory (so does the config one)

jagerman replied:
They are all copied from an original incomplete C destruction that does indeed leak.  I've pushed a commit here to fix it (both here, and two other places).  There's another one (for the network object) in the libquic API that I'll push there.

mpretty-cyro replied:
Thanks for pushing up a fix for these, and good catch @simophin 🙏

Bilb commented on include/session/config/groups/members.hpp:
```diff
@@ -127,6 +129,17 @@ struct member {
 
+    /// API: groups/member::invite_needs_send
+    ///
+    /// Returns whether the user needs an invite sent to them.  Returns true if so (whether or
+    /// not that invitation has been sent).
+    ///
+    /// Inputs: none
+    ///
+    /// Outputs:
+    /// - `bool` -- true if the user needs an invitation to be sent, false otherwise.
+    bool invite_needs_send() const { return invite_status == INVITE_NOT_SENT; }
```
should this maybe be tweak to make it clear that it isn't returning true if the invite was already tried to be sent, but failed?

mpretty-cyro replied:
Yea probably a good idea - I'll rename the functions to just be `invite_not_sent` and `promotion_not_sent` to avoid the ambiguity for now, I think down the track it'd be nicer to expose it as an enum instead of having individual functions but I wanted to avoid changing the behaviour too much at this stage

Bilb commented on include/session/config/groups/members.hpp:
```diff
@@ -155,20 +168,55 @@ struct member {
+        promotion_status = 0;
     }
 
+    /// API: groups/member::promotion_needs_send
+    ///
+    /// Returns whether the user needs a promotion to admin status to be sent.
+    /// Returns true if so (whether or not that promotion has failed).
+    ///
+    /// Inputs: None
+    ///
+    /// Outputs:
+    /// - `bool` -- true if the user needs a promotion to be sent, false otherwise.
+    bool promotion_needs_send() const { return promotion_status == INVITE_NOT_SENT; }
```
same comment, but this one might be a bit off with the right meaning too as this method wouldn't return true of `promotion_status == INVITE_FAILED` which is what `set_promotion_failed` does

mpretty-cyro replied:
Yea, I think this one is actually a bit more misleading than the other one...

Actually, looking at the logic I've changed the behaviour of the `admin` flag from "accepted admin" to be "someone that has been promoted to admin" - I'll update the functions accordingly (I think this behaviour makes more sense as we should consider a member an admin as soon as the promotion process has started, rather than only if they've confirmed receipt of the admin keys)

mpretty-cyro committed:
Renamed a couple of functions, updated docs, fixed 'promotion_x' funcs

jagerman committed:
Fix C free functions

These were meant to delete the internal C++ object, but weren't actually
doing so, and thus would be leaking the C++ objects when supposedly
being freed from the C API.

Also adds a missing `onion_request_builder_free` function and removes
the (wrong) doc implying that `onion_request_builder_build` frees the
object.

jagerman commented on include/session/onionreq/builder.h:
```diff
@@ -25,13 +25,21 @@ typedef struct onion_request_builder_object {
-/// When done with the object the `builder` must be destroyed by either passing the pointer to
-/// onion_request_builder_free() or onion_request_builder_build().
+/// When done with the object the `builder` must be destroyed by passing the pointer to
+/// onion_request_builder_free().
```
This comment was wrong (there was no builder_free function, the `_build()` function doesn't free anything).  And while _build *could* free, it seems a little weird to have it do so, because then you can't make one request builder and reuse it multiple times with different payloads.

The reason I comment though: if session-ios is calling this, it is worth double-checking that it isn't actually expecting `_build` to free the instance (and also adding a call to `_free` at the appropriate place).

mpretty-cyro replied:
There was a point where I'd updated session-ios to use these but before it got merged we ended up plugging `libQuic` in so now all the onion request building happens internally on the C++ side of things - so probably good to fix but luckily it won't affect anything currently 🎉

jagerman commented on include/session/config/groups/members.hpp:
```diff
@@ -79,14 +81,11 @@ struct member {
     /// Member variable
     ///
-    /// Flag that is set to indicate to the group that this member is an admin.
+    /// Flag that is set to indicate to the group that this member has been promoted to an admin.
```
This wording change seems slightly off, since the group creator never gets promoted to admin, and so this seems to be implying that it should be *false* for the creator because "has been promoted to" will be false.

jagerman replied:
Perhaps "is an admin or has an in-progress promotion to admin"

mpretty-cyro replied:
Ah yea, good catch - will update

jagerman commented on src/onionreq/builder.cpp:
```diff
@@ -252,7 +256,7 @@ LIBSESSION_C_API bool onion_request_builder_build(
 
     try {
-        auto unboxed_builder = unbox(builder);
+        auto& unboxed_builder = unbox(builder);
```
(This was unintentionally making a copy of the C++ builder object)

jagerman reviewed (approved):
LGTM (aside from the comments left).

mpretty-cyro committed:
Small docs change

mpretty-cyro committed:
Ran the formatter -_-
